### ISSUE TITLE
[FR] Updates for Klipper forks

### ISSUE THREAD

**Issue by naikymen** [Arksine/moonraker#615]
### Is your feature request related to a problem? Please describe

Nope.

### Describe the solution you'd like

I'd like to have Moonraker update my lovely Klipper fork for me, without having to SSH every time.

The most straightforward way to do this would be through an update_manager entry in the config file, but that seems impossible for now (read on).

### Describe alternatives you've considered

I have tried to setup the update_manager config for klipper, but that fails in the log with the following message:

```
2023-02-18 01:02:31,724 [moonraker.py:add_warning()] - [update_manager]: Extension klipper already added
```

I've patched some hard-coded stuff in Moonraker to get it working. It apparently used to disagree with my repo's origin, and it also didn't like my branch's name.

```diff
diff --git a/moonraker/components/simplyprint.py b/moonraker/components/simplyprint.py
index a3bfac3..ba33057 100644
--- a/moonraker/components/simplyprint.py
+++ b/moonraker/components/simplyprint.py
@@ -1002,7 +1002,8 @@ class SimplyPrint(Subscribable):
             "firmware": "Klipper",
             "firmware_version": version,
             "firmware_date": firmware_date,
-            "firmware_link": "https://github.com/Klipper3d/klipper",
+            # "firmware_link": "https://github.com/Klipper3d/klipper",
+            "firmware_link": "https://github.com/naikymen/klipper-homing-extruder",
         }
         diff = self._get_object_diff(fw_info, self.cache.firmware_info)
         if diff:
diff --git a/moonraker/components/update_manager/base_config.py b/moonraker/components/update_manager/base_config.py
index 4bcccbc..7c48aef 100644
--- a/moonraker/components/update_manager/base_config.py
+++ b/moonraker/components/update_manager/base_config.py
@@ -34,7 +34,9 @@ BASE_CONFIG: Dict[str, Dict[str, str]] = {
     },
     "klipper": {
         "moved_origin": "https://github.com/kevinoconnor/klipper.git",
-        "origin": "https://github.com/Klipper3d/klipper.git",
+        # "origin": "https://github.com/Klipper3d/klipper.git",
+        "origin": "https://github.com/naikymen/klipper-homing-extruder.git",
+        "primary_branch": "pipetting",
         "requirements": "scripts/klippy-requirements.txt",
         "venv_args": "-p python2",
         "install_script": "scripts/install-octopi.sh",
```

However this produces another issue, because moonraker is now "unclean", and cannot be updated.

![moonraker_pipetting](https://user-images.githubusercontent.com/3259326/219842905-f59b75cc-8198-43b4-ae9c-4ec25a7526fa.png)

### Additional information

There are over 4000 Klipper forks that might benefit from this feature.

Thanks!

**Comment by naikymen** [comment]
Possibly related: https://github.com/Arksine/moonraker/issues/470

**Comment by Arksine** [comment]
This is by design.  For security purposes it is necessary to make sure that sources are pristine from official repos.

You can use forks by [enabling debug features](https://moonraker.readthedocs.io/en/latest/installation/#command-line-usage) if you would like.   Keep in mind that your environment should be 100% secure/isolated if you do choose to do this, as this enables several features including the ability to view/modify the entire database.

**Comment by naikymen** [comment]
Thanks Arksine.

I am maintaining a Klipper fork for my project, which makes it the official fork for my machines.

Is forking moonraker the only alternative to have a working update manager for Klipper forks?

**Comment by charlespick** [comment]
> For security purposes it is necessary to make sure that sources are pristine from official repos.

@Arksine can you clarify what security purposes these are? The way I see it there should be a way to add a klipper fork to the update manager just like anything else.

**Comment by Arksine** [comment]
> can you clarify what security purposes these are? The way I see it there should be a way to add a klipper fork to the update manager just like anything else.

Most software configured via the update manager either depends on Moonraker or supports the printer in some way externally.  Klipper is a effectively a dependency for Moonraker, this makes it unique relative to other software in the ecosystem.  It isn't possible to predict what has changed in a fork, unpredictability is a potential security vulnerability.  Additionally, it can't be guaranteed that Moonraker will operate correctly with a forked version of Klipper.

That isn't to say that Moonraker prohibits the use of forks, only that Moonraker and its extensions do not have any kind of explicit support for them.   As previously mentioned, it is possible to update forks by enabling the debug option.  This provides an opt-in for developers and power users that want to run their own code.  The implication is that the user knows what they are doing and will take the steps necessary to mitigate risk.

> Q1: Is forking moonraker the only alternative to have a working update manager for Klipper forks? (without introducing vulnerabilities through debug mode).

As things currently stand those are your only options.  There are ways I could add support for updating forks that require an opt-in without enabling the debug option, and I empathize with the desire to do so.  Its a question of how I would go about it while being clear that Moonraker (or at least this repo) can't be expected to field bug reports or feature requests specific to forks.  I have given it thought and I best conclusion I can reach is to keep the policy as it is.

> Q2:  What simple change could be made to have update manager load a custom klipper origin from the moonraker.cfg file?

While I don't think it would be particularly difficult, I can't provide a "simple" change to do this off the top of my head.   The UM would need to fetch the configuration from the config object rather than `base_config.py`.  That said, Klipper and Moonraker have some unique aspects in the UM code, its possible that doing this would impact something else.

**Comment by naikymen** [comment]
Thanks Arksine! I ended up doing what we discussed.

The changes can be inspected here: https://github.com/Arksine/moonraker/compare/master...naikymen:moonraker-pipetting:pipetting#files_bucket

Briefly, the base_config dict is updated based on the values on the update_manager entry for Klipper.

There is also a fix for a possible bug in shell_command: https://github.com/naikymen/moonraker-pipetting/commit/1d61a439c927e93939f738cbd5e3f2b1270004ec

The update mechanism works, so I'm happy now :) Thanks again!

**Comment by Jookia** [comment]
For some data: I help run a fork of Klipper for printers that upstream won't add support for. The message shown is pretty scary: Klipper is invalid, click to recover. Clicking recover will break support for their printer. Rather than forking Moonraker I'm going to suggest enabling the debug mode.

**Comment by charlespick** [comment]
> I'd be interested to know what vulnerabilites the debug flag brings in and why.
> […](#)
> On Sun, May 14, 2023 at 12:39:11PM -0700, Nicolás A. Méndez wrote: The messages are scary indeed, but also very easy to get rid of (see diff above and links to my forks). Personally, I'd rather just trust myself and fork it, than introducing vulnerabilities. ![image](https://github.com/Arksine/moonraker/assets/3259326/597d53d3-5a2a-4ded-bb49-7dd7fe12792b) -- Reply to this email directly or view it on GitHub: [#615 (comment)](https://github.com/Arksine/moonraker/issues/615#issuecomment-1546985111) You are receiving this because you commented. Message ID: ***@***.***>

Arksine made an explanation of the vulnerabilities as they see them above. I don't really agree with it but it is what it is. Open source means if you don't like it then just make your own that doesn't toss up warnings about tiny things while leaving massive security holes elsewhere.

**Comment by naikymen** [comment]
Agreed, but that idea by itself also leads to fragmentation and duplicate efforts.

I just get the feeling that the klipper-software-things could be more flexible on the development side (e.g. open to extensions), but for some reason are not.

And being free to modify is not the same as being able to do so. For example, I now face the issue that Moonraker does not recognize git sumbodules as valid repositories. The code for this "validation" is really not straightforward, and I also do not understand the need for it.

The benefits of freedom outweigh my concerns on the possibility of eventual security issues, specially when no threat has been identified. Just look at Arch's AUR.

I insist on this only because I feel that giving people an option is always a good idea.

**Comment by Jookia** [comment]
> Open source means if you don't like it then just make your own that doesn't toss up warnings about tiny things while leaving massive security holes elsewhere.

Just to be clear, the massive security holes introduced here are not due to using a third-party Klipper but using the debug flag. This could be fixed by splitting up the debug flags like they were before.

**Comment by charlespick** [comment]
> > Open source means if you don't like it then just make your own that doesn't toss up warnings about tiny things while leaving massive security holes elsewhere.
> 
> Just to be clear, the massive security holes introduced here are not due to using a third-party Klipper but using the debug flag. This could be fixed by splitting up the debug flags like they were before.

There have been other security holes not related to the debug flag before. Not sure if they are still around as I haven't looked into it recently.

**Comment by dans98** [comment]
Id vote for the option to allow users to configure what repo they want to use.

**Comment by CroXY3D** [comment]
+1 to dans98, especially with the extreme gatekeeping of the main klipper repo to necessary fixes and features.

**Comment by Arksine** [comment]
FYI, Moonraker can now update forks of Klipper (or Moonraker) without any additional configuration and without enabling the debug flag.  Moonraker logs unofficial forks and provides that information to the front end to display, they can chose how and if to display it.  

>There have been other security holes not related to the debug flag before. Not sure if they are still around as I haven't looked into it recently.

Is that so?  I don't recall receiving reports of such.

**Comment by charlespick** [comment]
> Is that so? I don't recall receiving reports of such.

#449

**Comment by dans98** [comment]
> FYI, Moonraker can now update forks of Klipper (or Moonraker) without any additional configuration and without enabling the debug flag. Moonraker logs unofficial forks and provides that information to the front end, they can chose how and if to display it.
> 
> The primary security issue I had to overcome with supporting forks is that it would have opened an exploit using the recovery endpoint. I found a way to close that down.
> 


@Arksine maybe I've missed it, but is there documentation on changing/setting the repo?

**Comment by Arksine** [comment]
There is nothing to change from Moonraker's perspective.  Clone your fork, have Moonraker connect to it will detect its location.  From there it will initialize the repo, the primary difference being it will allow a mismatched remote and/or primary branch without setting the repo invalid.

Alternatively you can add another remote to an existing Klipper repo, fetch it and either have the master branch track it or switch to another branch.

**Comment by Arksine** [comment]
FYI:  I found and fixed a bug for the following scenario:
- You have the official version at `~/klipper`
- You clone a fork to another location, lets say `~/klipper-alt`
- Your modify the service to launch `klippy.py` to launch from the cloned location.

After restarting Klipper there was a bug that prevented the `update_manger` from instantiating the correct Deployer.  A restart of Moonraker would correct it, however I have just pushed a fix corrects instantiation in the above scenario.
